readthedocs/search/api.py

```diff
@@ -60,11 +62,50 @@ def get_inner_hits(self, obj):
 
 class PageSearchAPIView(generics.ListAPIView):
 
-    """Main entry point to perform a search using Elasticsearch."""
+    """
+    Main entry point to perform a search using Elasticsearch.
+
+    Required query params:
+    - q (search term)
+    - project
+    - version
+
+    .. note::
+
+       The methods `_get_project` and `_get_version`
+       are called many times, so a basic cache is implemented.
+    """
 
+    permission_classes = [IsAuthorizedToViewVersion]
     pagination_class = SearchPagination
     serializer_class = PageSearchSerializer
 
+    def _get_project(self):
+        cache_key = '_cached_project'
+        project = getattr(self, cache_key, None)
+
+        if not project:
+            project_slug = self.request.GET.get('project', None)
+            project = get_object_or_404(Project, slug=project_slug)
+            setattr(self, cache_key, project)
+
+        return project
+
+    def _get_version(self):
+        cache_key = '_cached_version'
+        version = getattr(self, cache_key, None)
+
+        if not version:
+            version_slug = self.request.GET.get('version', None)
+            project = self._get_project()
+            version = get_object_or_404(
+                project.versions.all(),
+                slug=version_slug,
+            )
+            setattr(self, cache_key, version)
+
+        return version
```

Comment:
Why do we need caching here?

Reply:
It's called 3 times, so we don't do the same query 3 times, there is a comment on the docstring from the class